lib/src/utils/async_queue.dart

```diff
@@ -0,0 +1,30 @@
+import 'dart:async';
+
+/// A queue that executes actions sequentially.
+///
+/// This is useful to avoid race conditions when accessing shared resources.
+class AsyncQueue<T> {
+  Completer<T>? _completer;
+
+  Future<T> execute(Future<T> Function() action) async {
+    while (_completer != null) {
+      try {
+        await _completer!.future;
+      } catch (_) {
+        // Ignores caught errors because it is handled below
+      }
+    }
+    try {
+      _completer = Completer<T>();
+
+      final result = await action();
+      _completer!.complete(result);
+      return result;
+    } catch (error, stackTrace) {
```

Comment:
We don't need to catch this error either. Let's just move
```
`_completer?.complete();`
`_completer = null;`
```
to the `finally` block